components/feature/customtabs/src/main/java/mozilla/components/feature/customtabs/verify/OriginVerifier.kt

```diff
@@ -0,0 +1,162 @@
+ * lifecycle and reuses that without making any new network requests.
+ *
+ * The lifecycle of this object is governed by the owner. The owner has to call
+ * {@link OriginVerifier#cleanUp()} for proper cleanup of dependencies.
+ */
+class OriginVerifier(
+    private val packageName: String,
+    @CustomTabsService.Relation private val relation: Int,
+    packageManager: PackageManager,
+    httpClient: Client
+) {
+
+    @VisibleForTesting
+    internal val handler = DigitalAssetLinksHandler(httpClient)
+    @VisibleForTesting
+    internal val signatureFingerprint by lazy {
+        getCertificateSHA256FingerprintForPackage(packageManager)
+    }
+
+    /**
+     * Verify the claimed origin for the cached package name asynchronously. This will end up
+     * making a network request for non-cached origins with a HTTP [Client].
+     *
+     * @param origin The postMessage origin the application is claiming to have. Can't be null.
+     */
+    suspend fun verifyOrigin(origin: Uri) = withContext(IO) { verifyOriginInternal(origin) }
+
+    @Suppress("LongMethod", "ReturnCount")
+    private fun verifyOriginInternal(origin: Uri): Boolean {
+        val cachedOrigin = cachedOriginMap[packageName]
+        if (cachedOrigin == origin) return true
+
+        if (origin.scheme != "https") return false
+        val relationship = when (relation) {
+            RELATION_USE_AS_ORIGIN -> USE_AS_ORIGIN
+            RELATION_HANDLE_ALL_URLS -> HANDLE_ALL_URLS
+            else -> return false
+        }
+
+        val originVerified = handler.checkDigitalAssetLinkRelationship(
+            origin.toString(),
+            packageName,
+            signatureFingerprint,
+            relationship = relationship
+        )
+
+        if (originVerified && packageName !in cachedOriginMap) {
+            cachedOriginMap[packageName] = origin
+        }
+        return originVerified
+    }
+
+    /**
+    * Computes the SHA256 certificate for the given package name. The app with the given package
+    * name has to be installed on device. The output will be a 30 long HEX string with : between
+    * each value.
+    * @return The SHA256 certificate for the package name.
+    */
+    @VisibleForTesting
+    internal fun getCertificateSHA256FingerprintForPackage(packageManager: PackageManager): String? {
+        val signature = packageManager.getSignature(packageName) ?: return null
+
+        val input = ByteArrayInputStream(signature.toByteArray())
+        return try {
+            val certificate = CertificateFactory.getInstance("X509").generateCertificate(input) as X509Certificate
+            byteArrayToHexString(MessageDigest.getInstance("SHA256").digest(certificate.encoded))
+        } catch (e: CertificateEncodingException) {
+            // Certificate type X509 encoding failed
+            null
+        } catch (e: CertificateException) {
+            // This shouldn't happen.
```

Comment:
If you are really sure that should not happen then I suggest throwing `AssertionError("Should not happen", e)` to verify your assumptions.